### PR TITLE
feat: Use macros for compiler id and version

### LINKED ISSUE #2935: Refactoring out compiler_info from build_info
@remyabel did a nice job on the spanner repo in https://github.com/googleapis/google-cloud-cpp-spanner/pull/298

Maybe I'll work on this unless @remyabel already started.

This is blocking https://github.com/googleapis/google-cloud-cpp-spanner/issues/21

### PR DESCRIPTION
**Pull request by tmatsuo**
fixes #2935

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2937)

### CODE REVIEW

File: google/cloud/internal/compiler_info_test.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+TEST(CompilerInfo, CompilerVersion) {
+  auto cv = CompilerVersion();
+  EXPECT_FALSE(cv.empty());
+#ifndef _WIN32  // gMock's regex brackets don't work on Windows.
+  // Look for something that looks vaguely like an X.Y version number.
+  EXPECT_THAT(cv, ::testing::ContainsRegex(R"([0-9]+.[0-9]+)"));
+#else
+  // Do our best on windows
+  EXPECT_EQ(std::string::npos, cv.find_first_not_of("01234567890."));
+  EXPECT_EQ(0, cv.find_first_of("01234567890"));
```
Comment (remyabel) [inline google/cloud/internal/compiler_info_test.cc]: Nit: you have 0 listed twice.

File: google/cloud/internal/compiler_info.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/compiler_info.h"
+//#include <algorithm>
+//#include <cctype>
+//#include <iterator>
```
Comment (remyabel) [inline google/cloud/internal/compiler_info.cc]: Nit: why are these commented headers here? Just remove them.

File: google/cloud/internal/compiler_info_test.cc (hunk 1 above)
Comment (tmatsuo) [reply google/cloud/internal/compiler_info_test.cc]: Good catch, removed

File: google/cloud/internal/compiler_info.cc (hunk 2 above)
Comment (tmatsuo) [reply google/cloud/internal/compiler_info.cc]: Done

**Pull request merged**